src/server/lock_manager/deadlock.rs

```diff
@@ -849,6 +1202,51 @@ where
         false
     }
 
+    /// Send `replace_locks_by_keys` request to the leader.
+    ///
+    /// Returns `Ok` if the request is sent to the client's channel. Otherwise,
+    /// the `items` will be returned back in the `Err`, so that the caller
+    /// can retry sending without cloning `items`.
+    fn send_replace_locks_by_keys_to_leader(
+        &mut self,
+        items: Vec<ReplaceLockByKeyItem>,
+    ) -> std::result::Result<(), Vec<ReplaceLockByKeyItem>> {
+        assert!(!self.is_leader() && self.leader_info.is_some());
+
+        if self.leader_client.is_none() {
+            self.reconnect_leader();
+        }
+        if let Some(leader_client) = &self.leader_client {
+            let mut replace_locks_by_keys_req = ReplaceLocksByKeysRequest::default();
+            replace_locks_by_keys_req.set_items(items.into());
+
+            let mut req = DeadlockRequest::default();
+            req.set_replace_locks_by_keys(replace_locks_by_keys_req);
+
+            // For old versions that does not support `replace_locks_by_keys`,
+            // it will recognize `tp` and `entry`. We let it tries to delete
+            // an entry that must not exist, so that it's a noop.
+            req.set_tp(DeadlockRequestType::CleanUpWaitFor);
+            let mut dummy_entry = WaitForEntry::default();
+            dummy_entry.set_txn(1);
+            dummy_entry.set_wait_for_txn(1);
+            req.set_entry(dummy_entry);
```

Comment:
What would happen without this?

Reply:
And if the leader is on an old version during the upgrade process, will the original `CleanUpWaitFor` logic be affected?